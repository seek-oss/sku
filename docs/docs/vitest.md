# Vitest

[Vitest] is a testing framework that supports ESM out-of-the-box, integrates with the Vite ecosystem and has a similar API to Jest.
These features make it a great replacement for Jest in `sku` applications, especially given [Jest's current limitations with ESM].
Sku supports Vitest as an alternate to Jest since v15

**We recommend using vitest with the vite bundler and in an ESM project**, however these are not requirements.

## Migrating to Vitest

Migrating to Vitest should be fairly straightforward.
Before making any changes, it is highly recommend to read [the Vitest documentation](https://vitest.dev/guide/) to familiarise yourself with the API and features, as well as to better understand the differences between Jest and Vitest.

To enable Vitest in `sku`, first install the required dependencies:

```bash
pnpm add -D vitest @sku-lib/vitest
```

Then, configure [`testRunner`][test runner] in your `sku` config:

```typescript
// sku.config.ts
import type { SkuConfig } from 'sku';

export default {
  testRunner: 'vitest',
  ...
} satisfies SkuConfig;
```

`sku` will now invoke [the `vitest` CLI][Vitest CLI] instead of the `jest` CLI when running `sku test`.

#### Key differences between Vitest and Jest

Vitest has strong compatibility with the Jest API, however it still has differences that may affect your tests.
Differences that affect sku projects are listed here.
The full list is documented in the Vitest documentation: [Migrating from Jest to Vitest].

To automate most of the migration process, a codemod is available.
Note that additional changes may still be required after running this codemod.

```sh
pnpm dlx @sku-lib/codemod jest-to-vitest .
```

**Default Watch Mode**

`vitest` defaults to watch mode when running tests.
To run tests without watch mode you can use the `--run` flag:

```json
// package.json
{
  "scripts": {
    // ...
    "test": "sku test --run"
  }
}
```

Watch mode won't trigger in CI environments, so it's safe to omit the flag in your pipeline.

**Code Coverage Dependencies**

Vitest does not install code coverage dependencies by default.
To collect code coverage, pass the `--coverage` flag to the `sku test` command and install the `@vitest/coverage-v8` package:

```bash
pnpm add -D @vitest/coverage-v8
```

**Testing Library Matchers**

If your test setup file includes an import for `@testing-library/jest-dom`, you may need to change this to `@testing-library/jest-dom/vitest`:

```diff
// test-setup.ts
- import '@testing-library/jest-dom';
+ import '@testing-library/jest-dom/vitest';
```

**Globals Disabled**

Jest enables global APIs such as `it`, `describe`, `beforeAll`, etc., by default.
Vitest does not.
Sku adopts the Vitest defaults, meaning you will need to explicitly import these test functions.
This change results in a cleaner global namespace, better type safety, clearer dependencies and is more consistency with modern javascript practices.

```diff
+ import { describe, expect, it } from 'vitest';
```

Be aware that since globals are disabled, some common libraries like `testing-library` will not run auto DOM cleanup.
If using these libraries, you will need to add cleanup to your configured `setupTests` file.

```diff
import '@testing-library/jest-dom/vitest';

+ import { cleanup } from '@testing-library/react';
+ import { afterEach } from 'vitest';

+ afterEach(cleanup)
```

[Vitest]: https://vitest.dev/
[Jest's current limitations with ESM]: https://jestjs.io/docs/ecmascript-modules
[test runner]: ./docs/configuration.md#testrunner
[Vitest CLI]: https://vitest.dev/guide/cli.html
[codemod]: https://codemod.com/registry/jest-vitest
[Migrating from Jest to Vitest]: https://vitest.dev/guide/migration.html#jest
[Vitest globals]: https://vitest.dev/config/#globals
