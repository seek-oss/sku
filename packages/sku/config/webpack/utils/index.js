const loaders = require('./loaders');
const { resolvePackage } = require('./resolvePackage');

module.exports = {
  ...loaders,
  resolvePackage,
  TYPESCRIPT: /\.(ts|tsx)$/,
  JAVASCRIPT: /\.js$/,
  LESS: /\.less$/,
  IMAGE: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
  SVG: /\.svg$/,
};
