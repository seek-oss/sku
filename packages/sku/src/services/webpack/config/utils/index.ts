import { extensions } from '../../../../config/fileResolutionExtensions.js';

export * from './loaders.js';
export { resolvePackage } from './resolvePackage.js';

export const TYPESCRIPT = new RegExp(`\.(${extensions.ts.join('|')})$`);
export const JAVASCRIPT = new RegExp(`\.(${extensions.js.join('|')})$`);
export const IMAGE = [
  /\.bmp$/,
  /\.gif$/,
  /\.jpe?g$/,
  /\.png$/,
  /\.webp$/,
  /\.avif$/,
];
export const SVG = /\.svg$/;
