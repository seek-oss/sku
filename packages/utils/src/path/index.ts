export { toPosixPath } from './toPosixPath.js';
