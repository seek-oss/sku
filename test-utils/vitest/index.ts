export { getAppSnapshot } from './appSnapshot.ts';
