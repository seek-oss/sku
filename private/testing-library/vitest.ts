// Re-export vitest setup from cli-testing-library
import 'cli-testing-library/vitest';
