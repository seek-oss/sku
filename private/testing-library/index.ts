import {
  render,
  configure,
  waitFor,
  cleanup,
  type Config,
  type RenderOptions,
} from 'cli-testing-library';
import { createRequire } from 'node:module';
import path from 'node:path';

const require = createRequire(import.meta.url);

type SkuCommand =
  | 'serve'
  | 'start'
  | 'start-ssr'
  | 'build'
  | 'build-ssr'
  | 'configure'
  | 'format'
  | 'init'
  | 'lint'
  | 'test';

// Default configuration for cli-testing-library
const DEFAULT_CONFIG: Partial<Config> = {
  asyncUtilTimeout: 10000,
};

// Configure cli-testing-library with default settings
configure(DEFAULT_CONFIG);

export const renderCli = async (
  command: SkuCommand,
  args: string[] = [],
  options: Partial<RenderOptions> = {},
) => render('node_modules/.bin/sku', [command, ...args], options);

export const configureCli = (config: Partial<typeof DEFAULT_CONFIG>) => {
  configure({ ...DEFAULT_CONFIG, ...config });
};

export const scopeToFixture = (fixtureFolder: string) => {
  const appDir = path.dirname(
    require.resolve(`../../fixtures/${fixtureFolder}/package.json`),
  );

  return {
    render: (
      command: SkuCommand,
      args: string[] = [],
      options: Partial<Omit<RenderOptions, 'cwd'>> = {},
    ) =>
      renderCli(command, args, {
        ...options,
        cwd: appDir,
      }),
  };
};

// Re-export the original functions for direct access
export { waitFor, cleanup };
