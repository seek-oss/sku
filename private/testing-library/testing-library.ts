import {
  render,
  configure,
  waitFor,
  cleanup as _cleanup,
  type Config,
  type RenderOptions,
} from 'cli-testing-library';
import { createRequire } from 'node:module';
import path from 'node:path';

const require = createRequire(import.meta.url);

type SkuCommand =
  | 'serve'
  | 'start'
  | 'start-ssr'
  | 'build'
  | 'build-ssr'
  | 'configure'
  | 'format'
  | 'init'
  | 'lint'
  | 'test';

// Default configuration for cli-testing-library
const DEFAULT_CONFIG: Partial<Config> = {
  asyncUtilTimeout: 5000,
};

// Configure cli-testing-library with default settings
configure(DEFAULT_CONFIG);

export const renderCli = async (
  command: SkuCommand,
  args: string[] = [],
  options: Partial<RenderOptions> = {},
) => render('node_modules/.bin/sku', [command, ...args], options);

export const configureCli = (config: Partial<typeof DEFAULT_CONFIG>) => {
  configure({ ...DEFAULT_CONFIG, ...config });
};

export const scopeToFixture = (fixtureFolder: string) => {
  const appDir = path.dirname(
    require.resolve(`../../fixtures/${fixtureFolder}/package.json`),
  );

  return {
    render: (
      command: SkuCommand,
      args: string[] = [],
      options: Partial<Omit<RenderOptions, 'cwd'>> = {},
    ) =>
      renderCli(command, args, {
        ...options,
        cwd: appDir,
      }),
    node: (
      args: string[] = [],
      options: Partial<Omit<RenderOptions, 'cwd'>> = {},
    ) =>
      render('node', args, {
        ...options,
        cwd: appDir,
      }),
  };
};

const skipCleanup = new Set<string>();

export const cleanup = async (task?: { id: string }) => {
  if (task?.id && skipCleanup.has(task.id)) {
    return;
  }

  await _cleanup();
};

export const deferCleanup = (
  id: string,
  onTestFinished: (fn: () => void) => void,
) => {
  skipCleanup.add(id);
  onTestFinished(() => {
    skipCleanup.delete(id);
  });
};

// Re-export the original functions for direct access
export { waitFor };
