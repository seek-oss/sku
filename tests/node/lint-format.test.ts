import { describe, it, beforeEach, afterEach, expect } from 'vitest';
import path from 'node:path';
import fs from 'node:fs/promises';
import dedent from 'dedent';
import { scopeToFixture } from '@sku-private/testing-library';

const { sku, fixturePath } = scopeToFixture('lint-format');

const srcDirectory = fixturePath('src');
const testFile = (fileName: string) => path.join(srcDirectory, fileName);

describe('lint-format', () => {
  beforeEach(async () => {
    await fs.mkdir(srcDirectory, { recursive: true });
  });

  afterEach(async () => {
    await fs.rm(srcDirectory, { recursive: true, force: true });
  });

  describe('sku lint', () => {
    it('should catch type errors', async () => {
      const fileName = 'typescriptFile.ts';
      const fileContents = dedent /* ts */ `
        const foo: number = 'a string';
      `;
      await fs.writeFile(testFile(fileName), fileContents);

      const lint = await sku('lint');
      expect(
        await lint.findByText('src/typescriptFile.ts(1,7): error'),
      ).toBeInTheConsole();
    });

    it('should catch ES lint errors', async () => {
      const fileName = 'utils.test.ts';
      const fileContents = dedent /* ts */ `
        console.log('foo');

        it.only('should test something', () => {
          let foo = true;

          expect(foo).toBe(true);
        });\n`;
      await fs.writeFile(testFile(fileName), fileContents);

      const lint = await sku('lint');
      expect(
        await lint.findByText('Checking code with ESLint'),
      ).toBeInTheConsole();
      expect(
        await lint.findByText('3 problems (3 errors, 0 warnings)'),
      ).toBeInTheConsole();
    });

    it('should catch prettier errors', async () => {
      const fileName = 'unformattedFile1.js';
      const fileContents = dedent /* js */ `
        import { something } from "with-double-quotes";

          const indented = 'indented';

        const foo = [
          "something \"really\" long",
          "that has to be moved",
          'to multiple lines',
          "so we can 'test' trailing commas"
        ];
      `;
      await fs.writeFile(testFile(fileName), fileContents);

      const lint = await sku('lint');
      expect(
        await lint.findByText('Checking code with Prettier'),
      ).toBeInTheConsole();
      expect(
        await lint.findByText('src/unformattedFile1.js'),
      ).toBeInTheConsole();
      expect(
        await lint.findByError(
          'Error: The file(s) listed above failed the prettier check',
        ),
      ).toBeInTheConsole();
      expect(
        await lint.findByText("To fix this issue, run 'pnpm run format'"),
      ).toBeInTheConsole();
    });
  });

  describe('sku format', () => {
    expect.addSnapshotSerializer({
      serialize: (val) => val,
      test: (val) => typeof val === 'string',
    });

    const filesToFormat: Record<string, string> = {
      'importOrder1.ts': dedent /* ts */ `
        import './reset'; // side-effect imports should stay put

        import LocalComponent from './LocalComponent'; // sibling
        import Parent from '../parent'; // parents
        import someModule from 'some-module'; // external
        import vanillaStyles from './vanillaStyles.css'; // styles
        import distantParent from '../../../parent'; // parents
        import myself from '.'; // index
        import path from 'path'; //  built-in
        import utils from 'src/utils'; // internal
      `,
      'importOrder2.ts': dedent /* ts */ `
        import aStyle from './a.css';
        import bStyle from './b.css';
        import cStyle from '../c.css';

        import b from './b';
        import a from './a';
        import c from '../c';
      `,
      'fixableLintError.ts': dedent /* ts */ `
        const foo = () => {
          return 'foo';
        };
      `,
      'unformattedFile2.ts': dedent /* ts */ `
        import { something } from "with-double-quotes";

          const indented = 'indented';

        const foo = [
          "something \"really\" long",
          "that has to be moved",
          'to multiple lines',
          "so we can 'test' trailing commas"
        ];
      `,
    };

    it.for(Object.keys(filesToFormat))(
      'errors are fixed: %s',
      async (fileName) => {
        const filePath = testFile(fileName);
        await fs.writeFile(filePath, filesToFormat[fileName]);

        const format = await sku('format');
        expect(
          await format.findByText('Formatting complete'),
        ).toBeInTheConsole();

        const result = await fs.readFile(filePath, { encoding: 'utf-8' });
        expect(result).toMatchSnapshot();
      },
    );
  });
});
