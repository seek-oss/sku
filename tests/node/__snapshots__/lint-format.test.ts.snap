// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lint-format > sku format > errors are fixed: fixableLintError.ts 1`] = `
const foo = () => 'foo';

`;

exports[`lint-format > sku format > errors are fixed: importOrder1.ts 1`] = `
import './reset'; // side-effect imports should stay put

import path from 'path'; //  built-in

import someModule from 'some-module'; // external

import utils from 'src/utils'; // internal

import distantParent from '../../../parent'; // parents
import Parent from '../parent'; // parents

import LocalComponent from './LocalComponent'; // sibling

import myself from '.'; // index

import vanillaStyles from './vanillaStyles.css'; // styles

`;

exports[`lint-format > sku format > errors are fixed: importOrder2.ts 1`] = `
import c from '../c';

import a from './a';
import b from './b';

import aStyle from './a.css';
import bStyle from './b.css';
import cStyle from '../c.css';

`;

exports[`lint-format > sku format > errors are fixed: unformattedFile2.ts 1`] = `
import { something } from 'with-double-quotes';

const indented = 'indented';

const foo = [
  'something "really" long',
  'that has to be moved',
  'to multiple lines',
  "so we can 'test' trailing commas",
];

`;
